### ISSUE TITLE
Thing Network Graph

### ISSUE THREAD

**Issue by rodneyrehm** [eclipse/smarthome#3492]
I'm a frontend developer at qivicon, where we've been dabbling (internally) with visualizing the ZigBee mesh network to better understand certain problems we have. We've realized that this sort of visualization could help our users better understand their own smart home networks. @kaikreuzer alerted me to similar (virtually identical) efforts @cdjackson has done, so here's my take on baking this into ESH itself.

## Example Graph

This graph is not meant to represent an actual rendering as we'd see in a UI. It's meant to help understand the data structure required to render something visually more appealing ;)

![Example Graph](https://rodneyrehm.de/t/network-schema.png)

In the lower right corner we see HomeMatic connected through an onboard chip. As far as I understand, most installations of ESH use the CCU or the network adapter, but that's beside the point. The HomeMatic network is a star topology with the coordinator in its center.

In the lower left corner we see the network (ethernet, wifi) connected devices. From our point of view ethernet provides a fix quality service, while WiFi devices may provide their current WiFi connection quality.

To the left we see devices that are connected via web services. Network and Internet are separate nodes in the graph to make that single-point-of-failure immediately obvious. (debatable, but I like the in-your-face style of this). The graph understands the connection hierarchy of the Netatmo devices because of the parent/bridge relationship ESH understands.

The Hue Bridge is a network connected device. The Hue bridge knows that the devices it provides communicate via ZigBee. However, the API lacks information about the actual network structure and signalStrengths of individual devices, so the structure is considered a star topology.

The ZigBee example in the upper right corner stands out for its mesh network. Unlike the other devices, each ZigBee device has connections to multiple other devices, with each link having its own signal quality. In this topology communication links are directional. The Coordinator may talk to Lamp 2 via Lamp 1, but Lamp 2 may respond via Lamp 3. (at least that's what I understood can happen)

## Data Structure

There are 2 entities; nodes and edges. The set of nodes consists of the ESH Things, a few generics (`root`, `network`, `internet`) and system specifics (e.g. HomeMatic chip, unless already represented by a Thing). However, constructing the set of edges is a bit more involved. Many connections can be inferred (given the right meta data), but some need to be provided by the binding (.e.g. the mesh network exemplified by ZigBee in the upper right corner).

```js
// Node
{
  // mandatory
  id: '<thing.id>',
  // mandatory
  label: '<thing.label>',
}

// Edge
{
  // mandatory
  from: '<id>',
  // mandatory
  to: '<id>',
  // [optional] type of connection
  // string, default null
  protocol: null,
  // [optional] wireless connection quality
  // integer, range [0, 4], default null
  signalStrength: null,
  // [optional] uni- or bi-directional connection
  // boolean, default false (bi-directional)
  directional: false,
}
```

For wireless connection quality (`edge.signalStrength`) I'd suggest we use the approach already established by the [signal-strength system channel](http://www.eclipse.org/smarthome/documentation/development/bindings/thing-definition.html#channels).

For the connection type (`edge.protocol`) Qivicon may have accidentally gotten a step ahead by requiring bindings to set the property `connection` on all bridge Things to one of the suggested values [`USB`, `SIMULATION`, `INTERNAL`, `INTERNET`, `NETWORK`]. We currently use this to make certain status messages a bit more explicit in qivicon UIs. I'd like to extend that ENUM by `RADIO` and introduce a new property `protocol`, that would take arbitrary strings ("ZigBee", "BidCos", "Z-Wave", …). Both properties default to `null`.

Usually connections are bi-directional, but especially in mesh networks we have uni-directional connections, each sporting their own `edge.signalStrength`. By default an edge is bi-directional (`edge.directional: false`) and can be downgraded where required.

### Example graph data

The above graph could be generated from the following data structure:

```js
{
  nodes: [
    { id: 'root', label: 'root' },
    { id: 'network', label: 'Network' },
    { id: 'internet', label: 'Internet' },

    { id: 'hmip', label: 'HomeMatic Coordinator' },
    { id: 'hmip:thermostat', label: 'Thermostat' },
    { id: 'hmip:watersensor', label: 'WaterSensor' },
    { id: 'hmip:motionsensor', label: 'MotionSensor' },

    { id: 'camera:1', label: 'IP Camera 1' },
    { id: 'camera:2', label: 'IP Camera 2' },

    { id: 'netatmo:service', label: 'Netatmo Service' },
    { id: 'netatmo:thermostat', label: 'Netatmo Thermostat' },
    { id: 'netatmo:module', label: 'Netatmo Module' },

    { id: 'hue:bridge', label: 'Hue Bridge' },
    { id: 'hue:a', label: 'Lamp A' },
    { id: 'hue:b', label: 'Lamp B' },

    { id: 'zigbee', label: 'ZigBee Coordinator' },
    { id: 'zigbee:1', label: 'Lamp 1' },
    { id: 'zigbee:2', label: 'Lamp 2' },
    { id: 'zigbee:3', label: 'Lamp 3' },
    { id: 'zigbee:4', label: 'Lamp 4' },
  ],
  edges: [
    { from: 'root', to: 'network', protocol: 'onboard' },
    { from: 'network', to: 'internet' },

    { from: 'root', to: 'hmip', protocol: 'onboard' },
    { from: 'hmip', to: 'hmip:thermostat', protocol: 'BidCos', signalStrength: 2 },
    { from: 'hmip', to: 'hmip:watersensor', protocol: 'BidCos', signalStrength: 4 },
    { from: 'hmip', to: 'hmip:motionsensor', protocol: 'HmIP', signalStrength: 3 },

    { from: 'network', to: 'camera:1' },
    { from: 'network', to: 'camera:2', protocol: 'wifi', signalStrength: 3 },

    { from: 'internet', to: 'netatmo:service' },
    { from: 'netatmo:service', to: 'netatmo:thermostat' },
    { from: 'netatmo:thermostat', to: 'netatmo:module' },

    { from: 'network', to: 'hue:bridge' },
    { from: 'hue:bridge', to: 'hue:a' },
    { from: 'hue:bridge', to: 'hue:b' },

    { from: 'root', to: 'zigbee', protocol: 'USB' },
    { from: 'zigbee', to: 'zigbee:1', protocol: 'ZigBee', signalStrength: 4, directional: true },
    { from: 'zigbee', to: 'zigbee:3', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:1', to: 'zigbee', protocol: 'ZigBee', signalStrength: 3, directional: true },
    { from: 'zigbee:1', to: 'zigbee:2', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:2', to: 'zigbee:1', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:2', to: 'zigbee:3', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:3', to: 'zigbee', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:3', to: 'zigbee:4', protocol: 'ZigBee', signalStrength: 2, directional: true },
    { from: 'zigbee:4', to: 'zigbee:2', protocol: 'ZigBee', signalStrength: 1, directional: true },
    { from: 'zigbee:4', to: 'zigbee:3', protocol: 'ZigBee', signalStrength: 2, directional: true },
  ]
}
```

## Customizations

In qivicon we still use the concept of rooms, as opposed to ESH's view of locations. I have therefore not included this information in the node object. However, I'd like to get that info right in the node as well, to avoid having to also load the list of things. It would be great if we could find a way to allow the solution (qivicon) to easily add some data, without having to overwrite the entire API.

## API

In the simplest approach above data could be obtained through `GET /rest/thing-network`. However, investigating networks like ZigBee can take some time (several 10s of seconds depending on the number of nodes), all the while blocking the HTTP request.

In qivicon we deal with these situations by responding from a cache and triggering the asynchronous/long-running process. The data, if not cached, is then sent in form of update events. This approach is a bit more involved, but would allow the UI to immediately render whatever is available and simply add nodes and edges as they come in (e.g. what you see when loading [this example](http://www.yasiv.com/youtube#?q=world)).

`GET /rest/thing-network` would return above structure from cache. If cache is empty, the response would be `{ nodes, edges: [] }` and the investigation of the edges in the network is initiated. Note that the nodes are already known, since they're basically the Things. The actual edges data will be sent via `smarthome/thing-network/added` events.

To make added|updated|removed events work for edges, they'd have to get an `id` property (which could just be the concatenation of `from` and `to`).

## Layout

Visually appealing graphs can be easily rendered by libraries like [vis.js](http://visjs.org/network_examples.html) or [d3](http://bl.ocks.org/paulovn/9686202). I'm likely going a bit lower level using [ngraph]() to achieve a mix of [this](http://www.yasiv.com/amazon#/Search?q=graph%20drawing&category=Books&lang=US) and [that](http://www.yasiv.com/youtube#?q=chrome). But the point is the positioning of the nodes and edges can be calculated.

### Floor Plans

Discussing this with @kaikreuzer we came back to the original idea of positioning nodes on top of a floor plan, similar to what [UniFi Network Management Controller](https://www.ubnt.com/software/) does:

![Network Heat Map](https://prd-www-cdn.ubnt.com/media/images/dashboard/right/UniFi/unifi-web-4.jpg)

However, I don't think we should be pursuing this idea just yet. It may be better for #582 to materialize a proper locations API and then map the locations to coordinates on floor plan instead. That might enable a sort of proximity relationship between things that could be useful beyond a visualization of the network.

## Protocol specifics

Certain protocols offer more data (and actions) than depicted in above graph. Since the available data varies, a generic interface isn't quite as simple. ZigBee, for example, allows resetting of the mesh network, dis/enabling node join, and provides the network address table.

We could abuse the infrastructure used to describe config parameters to allow bindings to set this stuff up. But I have too little insight to propose anything useful on this front at this point. Maybe @cdjackson has opinions?

Unless someone makes a good argument otherwise, I'd defer protocol specifics to a second iteration on the Thing Network API

## Investigating the network

I'm not particularly firm in the Java world, so I'll not insult you by attempting to propose an API for the bindings beyond explaining what I think should happen. When the network is investigated (either triggered by the REST API, or a scheduler), each binding should be asked to provide its edges. If a binding doesn't implement the hook, connections must be inferred by way of parent/bridge relationships and enhanced by connection/protocol properties and the signalStrength channel, if present. The hook should be asynchronous, because scanning the network may take some time.

**Comment by kaikreuzer** [comment]
@cdjackson As you have already experience with a similar feature in HABmin for Z-Wave, I would be very interested in your feedback on this idea!

**Comment by cdjackson** [comment]
Sorry - I missed this - it looks very interesting.

I have currently implemented a network graph "similar" to the first graph (I say similar - it's similar in structure, but looks different). This is implemented for the Zigbee binding under Paper UI (@kaikreuzer I think I sent you a picture of this running in PaperUI a couple of months back).

In my implementation I didn't want to mess with the base REST API due to the work I had at the time. 
I therefore exposed properties in the thing for each neighbor and each route - ie for each thing, there is a list of neighbors which contains the LQI and maybe some other stuff (eg joining flags), and a list of routes.

In the UI, I then generated the list of nodes and edges somewhat similar to what's presented above.

I'm a little sceptical about the idea of doing the node/edge translation on the server side. This defines an implementation to some degree (ie a network graph) - ok, I know that's the intention here, and maybe it's quite generic still, but I wonder if it's not better to stick with defining the neighbours, and possibly other properties, directly in the nodes and let the UI side derive the edges (or other implementation)? I know that's a bit more work in the UI, but it's a more generic interface and I'm worried that some UIs might need to generate this base information from the edges to get the neighbor list if they want to present the data in a different format.

Eg in my implementation, I allowed the user to click on a node, and in PaperUI this provided a pane on the right of the screen which displayed information like this nodes neighbours, and some other information. To get this we would need to traverse the edge list. If you then want to extend the information about neighbours (eg adding information that Zigbee provides such as the join status of the neighbour) this would either need to go in the edge, or we still need to define the neighbour into the node.

>Certain protocols offer more data (and actions) than depicted in above graph. Since the available data varies, a generic interface isn't quite as simple. ZigBee, for example, allows resetting of the mesh network, dis/enabling node join, and provides the network address table.

> We could abuse the infrastructure used to describe config parameters to allow bindings to set this stuff up. 

In my implementation this is exactly what I did to enable join or make a node leave.  It's the same as is already used for controller specific functions like network leave (at least in my bindings, and I think this is the agreed way as we discussed actions with Kai on the ESH forum a long time back). Such configuration could be grouped in a "network graph" parameter group. So, I don't think this is "abusing" the interface, although I still think the interface could be improved to better handle action/command type functionality...

The other thing that could be very useful here is the route list. To be clear - this is different to the neighbour list even if they are linked - ZigBee provides this information separately as a list of next_hop for a specific destination... This isn't something I implemented yet, but I had intended to have an option where the user can click on a node, and the route in use from the coordinator to that node would be highlighted by traversing the route list.

So, I would propose for discussion something like the following rather than nodes and edges... I think this is less implementation specific where network information is not used to directly present a graph, and it's also more extensible (IMHO).

```
{
  nodes: [
    {
      id: 'root'
      label: 'root'
      neighbors: [
        {
           destId: "node1",
           signalStrength: 234,
           joinStatus: true,
           routes: []
          etc...
        }
     ]
  ]
}
```

In the above, joinStatus and routes are protocol dependant information that could be added and used in an extended implementation.

> Note that the nodes are already known, since they're basically the Things

This isn't necessarily 100% true - it's only true if you add all the nodes as things. This is a limitation of the way I implemented the system and is potentially significant if you start a new network. I think it would be good to have a way to link nodes to things (ie provide the thing uid) so that the UI can render better information about the thing, but it should be optional in the event that a thing doesn't exist.

**Comment by cdjackson** [comment]
Sorry - I missed this - it looks very interesting.

I have currently implemented a network graph "similar" to the first graph (I say similar - it's similar in structure, but looks different). This is implemented for the Zigbee binding under Paper UI (@kaikreuzer I think I sent you a picture of this running in PaperUI a couple of months back).

In my implementation I didn't want to mess with the base REST API due to the work I had at the time. 
I therefore exposed properties in the thing for each neighbor and each route - ie for each thing, there is a list of neighbors which contains the LQI and maybe some other stuff (eg joining flags), and a list of routes.

In the UI, I then generated the list of nodes and edges somewhat similar to what's presented above.

I'm a little sceptical about the idea of doing the node/edge translation on the server side. This defines an implementation to some degree (ie a network graph) - ok, I know that's the intention here, and maybe it's quite generic still, but I wonder if it's not better to stick with defining the neighbours, and possibly other properties, directly in the nodes and let the UI side derive the edges (or other implementation)? I know that's a bit more work in the UI, but it's a more generic interface and I'm worried that some UIs might need to generate this base information from the edges to get the neighbor list if they want to present the data in a different format.

Eg in my implementation, I allowed the user to click on a node, and in PaperUI this provided a pane on the right of the screen which displayed information like this nodes neighbours, and some other information. To get this we would need to traverse the edge list. If you then want to extend the information about neighbours (eg adding information that Zigbee provides such as the join status of the neighbour) this would either need to go in the edge, or we still need to define the neighbour into the node.

>Certain protocols offer more data (and actions) than depicted in above graph. Since the available data varies, a generic interface isn't quite as simple. ZigBee, for example, allows resetting of the mesh network, dis/enabling node join, and provides the network address table.

> We could abuse the infrastructure used to describe config parameters to allow bindings to set this stuff up. 

In my implementation this is exactly what I did to enable join or make a node leave.  It's the same as is already used for controller specific functions like network leave (at least in my bindings, and I think this is the agreed way as we discussed actions with Kai on the ESH forum a long time back). Such configuration could be grouped in a "network graph" parameter group. So, I don't think this is "abusing" the interface, although I still think the interface could be improved to better handle action/command type functionality...

The other thing that could be very useful here is the route list. To be clear - this is different to the neighbour list even if they are linked - ZigBee provides this information separately as a list of next_hop for a specific destination... This isn't something I implemented yet, but I had intended to have an option where the user can click on a node, and the route in use from the coordinator to that node would be highlighted by traversing the route list.

So, I would propose for discussion something like the following rather than nodes and edges... I think this is less implementation specific where network information is not used to directly present a graph, and it's also more extensible (IMHO).

```
{
  nodes: [
    {
      id: 'root'
      label: 'root'
      neighbors: [
        {
           destId: "node1",
           signalStrength: 234,
           joinStatus: true,
           routes: []
          etc...
        }
     ]
  ]
}
```

In the above, joinStatus and routes are protocol dependant information that could be added and used in an extended implementation.

> Note that the nodes are already known, since they're basically the Things

This isn't necessarily 100% true - it's only true if you add all the nodes as things. This is a limitation of the way I implemented the system and is potentially significant if you start a new network. I think it would be good to have a way to link nodes to things (ie provide the thing uid) so that the UI can render better information about the thing, but it should be optional in the event that a thing doesn't exist.

**Comment by rodneyrehm** [comment]
> I'm a little sceptical about the idea of doing the node/edge translation on the server side.

I was looking for a (well defined) format that didn't require loading all the things and having to do any nodes/edges calculations in each client - this setup seemed like the obvious choice. However, I'm not fixated on this particular structure.


> but I wonder if it's not better to stick with defining the neighbours, and possibly other properties, directly in the nodes

What exactly would the list of neighbours include? Does each item in there contain the data defined for edges?

> The other thing that could be very useful here is the route list. 

Is this more than an array of IDs?


> > Note that the nodes are already known, since they're basically the Things
>
> This isn't necessarily 100% true - it's only true if you add all the nodes as things.

We already have non-Thing nodes (`root`, `network`, `internet`), but I had not considered networks that are only partially mapped to Things in ESH. That does complicate things to a degree where the edges returned by a binding would need to be checked against the known nodes in order to add these ghosts. If we stick with the nodes/edges design, the REST API would have to account for updates to nodes, too.

**Comment by cdjackson** [comment]
> I was looking for a (well defined) format that didn't require loading all the things and having to do any nodes/edges calculations in each client

Yes, but I guess you have the network graph in mind so defining nodes and edges makes sense. My point is that in my opinion this interface should provide the network topology such that it can be used in a graph, or other displays that may be of interest (you already showed a heatmap for example and I can think of others). These other displays would require the data in a different format rather than nodes and edges - I would propose to provide an interface that is generic and doesn't assume an implementation.  

I don't mean to download all the things though (ie through the thing interface) - the network interface would provide a consolidated list of nodes as you showed in your original proposal) just not the edges. Edges, or whatever the implementation needs should then be calculated from this (again - this part is dependent on the display to be produced though).

> What exactly would the list of neighbours include? Does each item in there contain the data defined for edges?

The neighbours list would simply define the neighbour and the signal strength - as I showed in my example earlier - it's the same information, just in a more consolidated format. It allows as well other neighbour information to be added if needed which would be difficult and inefficient if we only have edges.

> Is this more than an array of IDs?

It's an array or objects - for each node, there is a destination, and a next hop. Other information could also be added (maybe!) such as cost so I would keep it as an object.

**Comment by rodneyrehm** [comment]
> The neighbours list would simply define the neighbour and the signal strength - as I showed in my example earlier - it's the same information, just in a more consolidated format.

I'm good with changing the structure as long as I get all the data I need, including the protocol.

Would the `route` be populated for every node, or only where the binding knows what's going on? Is `joinStatus` a thing in multiple protocols, or is that a ZigBee only property?

**Comment by cdjackson** [comment]
I think these things are protocol dependant (joinStatus certainly is a ZigBee only property). I think we should have a way to define this sort of thing even if it’s not displayed in a “generic” UI - it gives the option to implement a specific UI for implementations that want to do this.  Maybe the best option is a dictionary of properties in the node (and potentially the neighbour as well?).

**Comment by rodneyrehm** [comment]
> Maybe the best option is a dictionary of properties in the node (and potentially the neighbour as well?).

Agreed. One node would be served up by one binding, so properties on a single node cannot clash. In the qivicon world we call this the `detail` map.

> I think we should have a way to define this sort of thing even if it’s not displayed in a “generic” UI - it gives the option to implement a specific UI for implementations that want to do this.

If these properties are described using standard means even a generic UI would have the chance to render *something*. Can we use standard config parameter descriptions for this?

The option to provide a "specific" UI for a binding is always there - but I doubt this will be done very often, if the "generic" UI can handle most, if not all of the binding specific data.

In the qivicon world we usually have one binding per protocol. How would we go about this to make sure that multiple bindings per protocol can provide the same fidelity?

**Comment by cdjackson** [comment]
> If these properties are described using standard means even a generic UI would have the chance to render something.

Yes, agreed it could render a list - same sort of thing as with thing properties now.

> Can we use standard config parameter descriptions for this?

I think my comment that you are responding to here was about adding protocol specific information - maybe you have a different view of this? I would have thought that a simple map, the same as the thing properties, would be fine. Why do you need the complexity of the parameter descriptions for a list of properties. There shouldn't be options, max/min limits, data types etc. I don't think the complexity of the config descriptions is needed?

> How would we go about this to make sure that multiple bindings per protocol can provide the same fidelity?

I guess you can't, and different protocols will have different information (and different fidelity) anyway. There should be a minimum that should be provided for "all" bindings, and after that I don't think you can ensure that all bindings provide the same information.

**Comment by rodneyrehm** [comment]
> I think my comment that you are responding to here was about adding protocol specific information - maybe you have a different view of this?

No, I agree with all of this being protocol specific. But protocol is not an entity in ESH so we can't really do anything with them right now. That's how I got back to bindings…

> There shouldn't be options, max/min limits, data types etc. I don't think the complexity of the config descriptions is needed?

I was thinking about (ab)using the config descriptions for their data type, format and labeling aspects. Maybe the Item metadata is more in line with what's needed here? From a generic UI perspective I care less about the semantic meaning of the data, more about the requirements to visualize it.

> > How would we go about this to make sure that multiple bindings per protocol can provide the same fidelity?
> 
> I guess you can't, and different protocols will have different information (and different fidelity) anyway.

I wasn't thinking about synchronizing the metadata of different protocols, but of different bindings for the *same* protocol. I assume some sort of documentation to communicate some sort of convention is the way to achieve this for now? We can, of course, defer this issue until we actually run into this problem.

**Comment by cdjackson** [comment]
> But protocol is not an entity in ESH so we can't really do anything with them right now. That's how I got back to bindings…

Ok, I was thinking in my mind protocol = binding. I know this isn't 100% true in some cases, but as you say we don't have any information on protocol, and to be honest I'm not really sure it's needed as most users probably don't care (or know) about the protocol.

**Comment by rodneyrehm** [comment]
> Ok, I was thinking in my mind protocol = binding. I know this isn't 100% true in some cases […]

This is certainly the case for everything qivicon does right now. But if I understood @kaikreuzer correctly, Bluetooth wouldn't be handled by a single binding. That said, ignoring the topic of aligning protocol specific data across bindings until the first real problem pops up is fine by me.

So far we've aligned on my data in your structure including a `detail` map for protocol specific data. But we haven't yet agreed on config descriptions or an alternative. If we get this "right" there should be very little incentive to create any custom UI, since the generic UI can handle the bulk. If we omit this, I doubt many bindings will supply any protocol specific data, since they'd also have to create an UI for that…
